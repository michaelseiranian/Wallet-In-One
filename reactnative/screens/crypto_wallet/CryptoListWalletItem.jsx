import {
  Image,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from "react-native";
import React, {useEffect, useState} from "react";
import getCryptoIcon from "../cryptocurrency/icons/icon";
import { useTheme } from '../../src/theme/ThemeProvider'

/**
 * Component that displays a crypto wallet within the crypto wallet list, which includes a logo, the balance of the
 * crypto wallet in its currency, and the value of the crypto wallet in pounds.
 */
export default function CryptoListWalletItem(props) {

  const [cryptoValue, setCryptoValue] = useState(0);
  const {dark, colors, setScheme} = useTheme();

  const styles = StyleSheet.create({
    walletAsset: {
      padding: 10,
      marginVertical: 5,
      borderRadius: 10,
      flexDirection: "row",
    },
    walletAssetTitle: {
      fontWeight: "700",
      flex: 1,
    },
    walletAssetImage: {
      width: 30,
      height: 30,
    },
  });

  /**
   * Function that gets the conversion rate of the target cryptocurrency into pounds and stores it in the component.
   */
  const getCryptoValue = async () => {
    await fetch(`https://min-api.cryptocompare.com/data/price?fsym=${props.item.symbol}&tsyms=GBP`)
      .then((res) => res.json())
      .then((res) => res.GBP)
      .then((res) => setCryptoValue(res))
      .catch((err) => console.log(err))
  }

  useEffect(() => {
    getCryptoValue();
  }, []);

  return (
    <TouchableOpacity
      onPress={() =>
        props.navigation.navigate("Crypto Wallet Detail",
          { id: props.item.id, value: cryptoValue * props.item.balance, removeWallet: props.removeWallet })
      }
    >
      <View style={[styles.walletAsset, {backgroundColor: colors.primary}]}>
        <View
          style={{
            alignItems: "center",
            justifyContent: "center",
            paddingRight: 10,
          }}
        >
          <Image
            style={styles.walletAssetImage}
            source={getCryptoIcon(props.item.symbol)}
          />
        </View>

        <View
          style={{
            flex: 1,
            flexDirection: "row",
            justifyContent: "space-between",
          }}
        >
          <View style={{}}>
            <Text style={{ fontSize: 25, fontWeight: "700", color: colors.text }}>
              {props.item.cryptocurrency}
            </Text>
            <Text style={[styles.walletAssetTitle, {color: colors.text}]}>
              {props.item.balance} {props.item.symbol}
            </Text>

            <Text style={[styles.walletAssetTitle, {color: colors.text}]}>Â£{(cryptoValue * props.item.balance).toFixed(2)}</Text>
          </View>

        </View>
      </View>
    </TouchableOpacity>
  );
}