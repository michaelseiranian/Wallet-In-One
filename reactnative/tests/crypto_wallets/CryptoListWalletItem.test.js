import {render, act, fireEvent, screen, waitFor} from "@testing-library/react-native";
import CryptoListWalletItem from "../../screens/crypto_wallet/CryptoListWalletItem";


describe('<CryptoListWalletItem />', () => {

  const navigate = jest.fn();

  beforeEach(() => {
    global.fetch =  jest.fn( async () => {
      return Promise.resolve({
        status: 200, json: () => ({'GBP': 5}),
      })
    })
  })

  afterEach(() => {
    global.fetch.mockClear();
    delete global.fetch;
  })

  it('snapshot test', async () => {
    const snapshot = render(<CryptoListWalletItem item={{symbol: '-', id: 1}} navigation={{navigate}} />)

    await act(async () => {
      expect(snapshot).toMatchSnapshot()

      // Test navigation
      fireEvent.press(await screen.getByText('-'))

    })
  })

  it('wallet data test', async () => {
    const snapshot = render(<CryptoListWalletItem item={{symbol: 'BTC', cryptocurrency: 'Bitcoin', balance: 10.01}} />)

    await act(async () => {

      expect(screen.getByText("Bitcoin"))
      expect(screen.getByText("10.01 BTC"))

      await waitFor(() => {
        expect(screen.getByText('£50.05')).toBeDefined()
      })

      expect(snapshot).toMatchSnapshot()

    })
  })

  it('fetch fail test', async() => {

    global.fetch =  jest.fn( async () => {
      return Promise.resolve({
        status: 200, json: () => Promise.reject(),
      })
    })

    const snapshot = render(<CryptoListWalletItem item={{symbol: 'BTC', cryptocurrency: 'Bitcoin', balance: 10.01}} />)

    await act(async () => {

      expect(screen.getByText("Bitcoin"))
      expect(screen.getByText("10.01 BTC"))

      await waitFor(() => {
        expect(screen.getByText('£0.00')).toBeDefined()
      })

      expect(snapshot).toMatchSnapshot()

    })
  })

})