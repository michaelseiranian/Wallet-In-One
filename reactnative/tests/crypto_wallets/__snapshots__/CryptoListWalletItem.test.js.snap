// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoListWalletItem /> fetch fail test 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "borderRadius": 10,
          "flexDirection": "row",
          "marginVertical": 5,
          "padding": 10,
        },
        {
          "backgroundColor": "purple",
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "paddingRight": 10,
        }
      }
    >
      <Image
        source={1}
        style={
          {
            "height": 30,
            "width": 30,
          }
        }
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={{}}
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 25,
              "fontWeight": "700",
            }
          }
        >
          Bitcoin
        </Text>
        <Text
          style={
            [
              {
                "flex": 1,
                "fontWeight": "700",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          10.01
           
          BTC
        </Text>
        <Text
          style={
            [
              {
                "flex": 1,
                "fontWeight": "700",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          £
          0.00
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<CryptoListWalletItem /> snapshot test 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "borderRadius": 10,
          "flexDirection": "row",
          "marginVertical": 5,
          "padding": 10,
        },
        {
          "backgroundColor": "purple",
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "paddingRight": 10,
        }
      }
    >
      <Image
        style={
          {
            "height": 30,
            "width": 30,
          }
        }
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={{}}
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 25,
              "fontWeight": "700",
            }
          }
        />
        <Text
          style={
            [
              {
                "flex": 1,
                "fontWeight": "700",
              },
              {
                "color": "black",
              },
            ]
          }
        >
           
          -
        </Text>
        <Text
          style={
            [
              {
                "flex": 1,
                "fontWeight": "700",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          £
          NaN
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<CryptoListWalletItem /> wallet data test 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "borderRadius": 10,
          "flexDirection": "row",
          "marginVertical": 5,
          "padding": 10,
        },
        {
          "backgroundColor": "purple",
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "paddingRight": 10,
        }
      }
    >
      <Image
        source={1}
        style={
          {
            "height": 30,
            "width": 30,
          }
        }
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={{}}
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 25,
              "fontWeight": "700",
            }
          }
        >
          Bitcoin
        </Text>
        <Text
          style={
            [
              {
                "flex": 1,
                "fontWeight": "700",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          10.01
           
          BTC
        </Text>
        <Text
          style={
            [
              {
                "flex": 1,
                "fontWeight": "700",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          £
          50.05
        </Text>
      </View>
    </View>
  </View>
</View>
`;
